packages/jupyter-ai-magics/jupyter_ai_magics/providers.py

```diff
@@ -475,7 +502,20 @@ async def _acall(self, *args, **kwargs) -> Coroutine[Any, Any, str]:
         return await self._call_in_executor(*args, **kwargs)
 
 
-class OpenAIProvider(BaseProvider, OpenAI):
+class OpenAIBaseProvider(BaseProvider):
+    """
+    Determine if the exception is an OpenAI API key error.
+    """
+
+    @classmethod
+    def is_api_key_exc(cls, e: Exception):
+        if isinstance(e, openai.error.AuthenticationError):
+            error_details = e.json_body.get("error", {})
+            return error_details.get("code") == "invalid_api_key"
+        return False
```

Comment:
This is not needed. We can move the code to a function, and call that inside the OpenAI providers. There is slight duplication in terms of declaring the exception method, but avoids the risk of creating another class hierarchy than the `BaseProvider`.